internal/ip/ip.go

```diff
@@ -0,0 +1,179 @@
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package ip
+
+import (
+	"fmt"
+	"net"
+	"syscall"
+
+	"github.com/uber/arachne/defines"
+	"github.com/uber/arachne/internal/log"
+
+	"github.com/google/gopacket"
+	"github.com/google/gopacket/layers"
+	"github.com/google/gopacket/pcap"
+	"go.uber.org/zap"
+)
+
+// Conn represents the underyling functionality to send and recv Arachne echo requests
+type Conn struct {
+	SrcAddr net.IP
+	AF      int
+	sendFD  int
+	recvSrc gopacket.PacketDataSource
+}
+
+// Close is used to close a Conn's send file descriptor and Recv packet source
+func (c *Conn) Close() error {
+	return syscall.Close(c.sendFD)
+}
+
+// NextPacket gets the next available packet from the PacketDataSource
+func (c *Conn) NextPacket() (gopacket.Packet, error) {
+	data, _, err := c.recvSrc.ReadPacketData()
+	if err != nil {
+		return nil, err
+	}
+	return gopacket.NewPacket(data, layers.LinkTypeEthernet, gopacket.DecodeOptions{Lazy: true}), nil
+}
+
+// Sendto operates on a Conn file descriptor and mirrors the Sendto syscall
+func (c *Conn) Sendto(b []byte, to net.IP) error {
+	sockAddr, err := ipToSockaddr(c.AF, to, 0)
+	if err != nil {
+		return err
+	}
+
+	return syscall.Sendto(c.sendFD, b, 0, sockAddr)
+}
+
+// getSendSocket will create a raw socket for sending data
+func getSendSocket(af int) (int, error) {
+	fd, err := syscall.Socket(af, syscall.SOCK_RAW, syscall.IPPROTO_RAW)
+	if err != nil {
+		return 0, err
+	}
+
+	err = syscall.SetsockoptInt(fd, syscall.IPPROTO_IP, syscall.IP_HDRINCL, 1)
+	if err != nil {
+		return 0, err
+	}
+
+	return fd, nil
+}
+
+func getRecvSource(listenPort uint32, intf string) (gopacket.PacketDataSource, error) {
+	// Filter for tcp packets coming into ListenPort, http, or HTTPS and contain a SYN flag
+	filter := fmt.Sprintf("(dst port %d or dst port %d or dst port %d) and (tcp[13] & 2 != 0)",
+		listenPort,
+		defines.PortHTTP,
+		defines.PortHTTPS)
+
+	handle, err := pcap.OpenLive(intf, defines.PcapMaxSnapLen, true, pcap.BlockForever)
+	if err != nil {
+		return nil, err
+	}
+
+	if err = handle.SetBPFFilter(filter); err != nil {
+		return nil, err
+	}
+
+	handle.SetDirection(pcap.DirectionIn)
+
+	return handle, nil
+}
+
+// NewConn returns a raw socket connection to send and receive packets
+func NewConn(af int, listenPort uint32, intf string, srcAddr net.IP, logger *log.Logger) *Conn {
+	fds, err := getSendSocket(af)
+	if err != nil {
+		logger.Fatal("Error creating send socket",
```

Comment:
fatal on anything but startup is an anti-pattern, but this repo may not subscribe to that.